### PR TITLE
Add basic constant folding rules to eggcc

### PR DESCRIPTION
**Pull request by oflatt**

### CODE REVIEW

File: src/rvsdg/egglog_optimizer/mod.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    "(run 3)".to_string()
+}
+
+#[derive(Debug, PartialEq, Clone)]
+struct BrilOp {
+    op: &'static str,
+    egglog_op: &'static str,
+    // so far we don't need more than 2 types
+    // if the field is None, it is beyond the number of inputs
+    input_types: [Option<Type>; 2],
```
Comment (yihozhang) [inline src/rvsdg/egglog_optimizer/mod.rs:26]: this looks a bit uncommon to me. Why not `Vec<Type>`?

File: src/rvsdg/egglog_optimizer/constant_fold.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+pub(crate) fn constant_fold_egglog() -> String {
+    let mut res = vec![];
+
+    for bril_op in BRIL_OPS {
+        let op = bril_op.op;
+        let egglog_op = bril_op.egglog_op;
+        if bril_op.num_inputs() == 2 {
+            res.push(format!(
+                "(rewrite ({op} ty
```
Comment (yihozhang) [inline src/rvsdg/egglog_optimizer/constant_fold.rs:11]: Alas if only we had metaprogramming in egglog.

**Review by yihozhang — APPROVED** [review]

File: src/rvsdg/egglog_optimizer/mod.rs (hunk 1 above)
Comment (oflatt) [reply src/rvsdg/egglog_optimizer/mod.rs:26]: I'm declaring a global type so it needs to have a fixed size apparently

File: src/rvsdg/egglog_optimizer/constant_fold.rs (hunk 2 above)
Comment (oflatt) [reply src/rvsdg/egglog_optimizer/constant_fold.rs:11]: alas

**Pull request merged**